Lib/test/test_importlib/identity.py

```diff
@@ -0,0 +1,521 @@
+    return text
+
+
+def get_name_qchar_idxs(text, succeeded):
+    """
+    Given any legal input (ie, not matching "[^ ]@."), yield the
+    character indexes for which the text remains valid when the
+    index is substituted with a quote character.
+    """
+    replace_at = re.compile(
+        rf"""
+        # Cannot replace " " before "@" if "@" succeeded by anything.
+        (?! $ | [ ]@. {"| [ ]@$" if succeeded else ""})
+        """,
+        re.VERBOSE,
+    )
+    return (m.start() for m in replace_at.finditer(text))
+
+
+@st.composite
+def identities_strategy(draw, simple_chars=False, debug_entries=False):
+    """
+    For each identity field of the core metadata specification ("Author",
+    "Maintainer", "Author-email" and "Maintainer-email"), generate a list
+    of ``Entry`` instances, each corresponding to a possibly-empty
+    identity. The "-email" fields have stricter requirements and generate
+    from ``ident_entry`` whereas the non-email fields use ``name_entry``.
+    Empty lists are dropped. Process and combine each non-empty list into
+    an unstructured string conforming to the core metadata specification
+    as well as into a more succinct structured format - two lists of
+    (name, email) tuples, one for authors and the other for maintainers.
+
+    If ``simple_chars``, restrict the list of allowed characters to
+    reduce the Hypothesis search space and permit easier visualization.
+    If ``debug_entries``, return the underlying lists of ``Entry``
+    instances in addition to the unstructured string and structured list
+    representations.
+
+    This strategy will generate the most permissible identities possible:
+    unbalanced quote characters, unescaped special characters, omitted
+    identity information, empty identities, and so forth.
+    """
+    text = ""
+    authors = set()
+    maintainers = set()
+    entry_dict = dict()
+    state = types.SimpleNamespace(simple_chars=simple_chars)
+
+    combinations = zip(
+        ((i, j) for i in (name_entry, ident_entry) for j in (authors, maintainers)),
+        ("Author", "Maintainer", "Author-email", "Maintainer-email"),
+    )
+
+    for (entry_type, entry_target), entry_prefix in combinations:
+        entries, data_set, data_str = draw(identity_entries(state, entry_type))
+        entry_dict[entry_prefix] = entries
+        if not entries:
+            continue
+        text += entry_prefix + ": " + data_str + "\n"
+        entry_target |= data_set
+
+    result = (text, authors, maintainers)
+    if debug_entries:
+        result = (entry_dict, *result)
+    return result
+
+
+@st.composite
+def identity_entries(draw, state, entry_strategy):
+    """
+    Generate a list of processed identity entries using
+    ``entry_strategy``.
+    """
+    sample = draw(st.lists(entry_strategy(state), min_size=0, max_size=10))
+    for entry in sample[:-1]:
+        entry.append(Token(identity_entries, ", ", True))
+    unbalance(sample)
```

Comment:
This is a no-op; did you mean
```suggestion
    draw(unbalance(sample))
```

I'm also generally uncomfortable with strategies which mutate their argument/s in place rather than generating a fresh value, but that's a question of idiom rather than function per se.

Reply:
Hmm drawing a mutation is indeed discomfiting - going to have to fix that.